src/network/networkd-netdev-wireguard.c

```diff
@@ -0,0 +1,625 @@
+/***
+    This file is part of systemd.
+
+    Copyright 2016 Jörg Thalheim <joerg@higgsboson.tk>
+
+    systemd is free software; you can redistribute it and/or modify it
+    under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    systemd is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <sys/ioctl.h>
+#include <net/if.h>
+
+#include "alloc-util.h"
+#include "parse-util.h"
+#include "in-addr-util.h"
+#include "fd-util.h"
+#include "strv.h"
+#include "hexdecoct.h"
+#include "string-util.h"
+#include "networkd-netdev-wireguard.h"
+#include "networkd-link.h"
+#include "networkd.h"
+
+#define WG_SET_DEVICE (SIOCDEVPRIVATE + 1)
+#define MAX_CONCURRENT_QUERIES 2048
+
+static void resolve_endpoints(NetDev *netdev);
+
+static WireguardPeer *new_peer(Wireguard *w, unsigned section) {
+        WireguardPeer *peer;
+
+        assert(w);
+
+        if (w->last_peer_section == section && w->peers)
+                return w->peers;
+
+        peer = new0(WireguardPeer, 1);
+        if (!peer)
+                return NULL;
+        peer->fields.replace_ipmasks = true;
+        peer->fields.persistent_keepalive_interval = (uint16_t)-1;
+
+        LIST_PREPEND(peers, w->peers, peer);
+        w->allocation_size += sizeof(struct wgpeer);
+        w->last_peer_section = section;
+        w->dev.num_peers++;
+
+        return peer;
+}
+
+static int update_wireguard_config(NetDev *netdev) {
+        _cleanup_close_ int fd = -2;
+        _cleanup_free_ void *data = NULL;
+        void *pos = NULL;
+        struct ifreq ifreq = {0};
+        size_t len;
+        Wireguard *w;
+        WireguardPeer *peer;
+        WireguardIpmask *mask;
+
+        assert(netdev);
+
+        w = WIREGUARD(netdev);
+
+        assert(w);
+
+        fd = socket(AF_INET, SOCK_DGRAM, 0);
```

Comment:
please always open fds with cloexec right away. i.e. add SOCK_CLOEXEC here to the second apram...